frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/model/ResourceDescriptor.java

```diff
@@ -290,6 +291,45 @@ public Resource toResource() throws Exception {
     return res;
   }
 
+  // format user request resource to hadoop yarn support resource.
+  // ignore Ports and GPUs when hadoop yarn library doesn't support these two features.
+  public static ResourceDescriptor formatResource(ResourceDescriptor resourceDescriptor) throws Exception {
+    ResourceDescriptor formatedResource = YamlUtils.deepCopy(resourceDescriptor, ResourceDescriptor.class);
+
+    Class<?> clazz = Resource.newInstance(0, 0).getClass();
+
+    if (formatedResource.getGpuNumber() > 0 || formatedResource.getGpuAttribute() > 0) {
+      try {
+        clazz.getMethod("setGPUs", int.class);
+        clazz.getMethod("setGPUAttribute", long.class);
+      } catch (NoSuchMethodException e) {
+        LOGGER.logWarning(e, "YARN library doesn't support GPU, ingore the GPU related user input");
+        formatedResource.setGpuNumber(0);
+        formatedResource.setGpuAttribute(0L);
+      }
+    }
+    if (formatedResource.getPortNumber() > 0 || ValueRangeUtils.getValueNumber(formatedResource.getPortRanges()) > 0) {
+      try {
+
+        Class hadoopValueRangesClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRanges");
+        Class hadoopValueRangeClass = Class.forName("org.apache.hadoop.yarn.api.records.ValueRange");
+
+        hadoopValueRangesClass.getMethod("newInstance").invoke(null);
+        hadoopValueRangesClass.getMethod("setRangesList", List.class);
+        clazz.getMethod("setPorts", hadoopValueRangesClass);
+      } catch (NoSuchMethodException e) {
+        LOGGER.logWarning(e, "YARN library doesn't support Port allocation, ignore port related user input");
+        formatedResource.setPortNumber(0);
+        formatedResource.setPortRanges(new ArrayList<>());
+      } catch (ClassNotFoundException e) {
+        LOGGER.logWarning(e, "YARN library doesn't support Port allocation, ignore port related user input");
+        formatedResource.setPortNumber(0);
+        formatedResource.setPortRanges(new ArrayList<>());
+      }
+    }
```

Comment:
Duplicated implement. Why not just:
return ResourceDescriptor.fromResource(resourceDescriptor.toResource())